csrc/expr_evaluator.cpp

```diff
@@ -149,6 +149,10 @@ void ExpressionEvaluator::bind_(
         // Ignore concrete extents because they hold the unsharded extents.
         NVF_CHECK(
             1 == t.size(i), "Tried to bind a constant value 1 as ", t.size(0));
+        bind_(
```

Comment:
I wonder why we need this for DIDx but not for TIDx/BIDx

Reply:
I am not sure to understand. Tid/Bid are not needed to compute the allocation size (at least in our examples)

The extent of the shared axis can only be deduced through the device mesh's size

Reply:
> The extent of the shared axis can only be deduced through the device mesh's size

Generally, this is not true. When a domain is annotated with, e.g., TIDx, it only means `blockDim.x` is greater than or equal to the extent of the domain. 

It also seems strange that the extent of a domain is determined by the number of devices. Shouldn't it be the other way around?

Reply:
> When a domain is annotated with, e.g., TIDx, it only means blockDim.x is greater than or equal to the extent of the domain.

If BlockDim.x > TIDx extent does this imply that some threads in will not allocate local memory/perform any computation related to that tensor?

For sharding, we want the constraint that if a tensor is sharded over a mesh, that at least some part of the tensor is allocated on those devices in the mesh. 
Then we can say if a tensor of size [1, 4] is sharded on dimension 0 across mesh{0,1,2}, we know the unsharded size must be [3,4]

Reply:
> It also seems strange that the extent of a domain is determined by the number of devices. Shouldn't it be the other way around?

We can see it in both way, the values just need to match. In our case since the DeviceMesh are not yet symbolic they hold the concrete value (which is not necessarily the number of devices)

I will soon make the DeviceMesh symbolic and those considerations will be made clearer hopefully

Reply:
> If BlockDim.x > TIDx extent does this imply that some threads in will not allocate local memory/perform any computation related to that tensor?

Yes

> For sharding, we want the constraint that if a tensor is sharded over a mesh, that at least some part of the tensor is allocated on those devices in the mesh.

Yes, having a constraint like this is fine. I just wanted to make sure you're aware of the deviation from TIDx/BIDx.

Reply:
> Generally, this is not true. When a domain is annotated with, e.g., TIDx, it only means `blockDim.x` is greater than or equal to the extent of the domain.

We only support the case where there is strict equality between the axis's and DeviceMesh's extent. I don't even know what it would mean if we have just an inequality (especially in our case where we forbid split/merge)

More fundamentally, the Fusion needs to know __somehow__ the extents of the sharded axis, which cannot be deduced by the tensor's concrete shape (since sharded dimensions are not allocated)

All these are important points but they are not related to this PR though :)

Reply:
> I will soon make the DeviceMesh symbolic

So, then how would you determine the extent?

Reply:
And back to my original question, I'm trying to understand why we need this. We don't need to infer tensor shapes from blockDim.x or gridDim.x, but somehow we need for DIDx parallelized domains. I'm not saying we should not make [this assumption](https://github.com/NVIDIA/Fuser/pull/1900#discussion_r1522139313), and I understand that the assumption allows us to do the inference of extents. My question is why it's necessary only for DIDx parallelized domains.

Reply:
Oh I see your question.
This needed when a global input is a sharded. Since the input to the program will be the sharded tensor, we need to calculate unsharded extents.

Reply:
Thanks @samnordmann , I think I now understand why.

> since sharded dimensions are not allocated

So, the fundamental issue is the input to nvFuser, i.e., an aten tensor, does not convey anything about the distributed allocation, so we need to make this assumption atm. Ideally, sharding information should be also given to nvFuser, which should also tell the overall logical shape.

Does this make sense?

Reply:
@samnordmann Could you please add some more comments to the code?

Reply:
Yep, that make sense. Let me add some comment